File: summary/ruby/22_annealing.rb
def simulated_annealing(initial_solution)
  current_solution = initial_solution.dup
  current_cost = yield(current_solution)
  temperature = 100.0
  min_temperature = 0.01
  alpha = 0.9

  while temperature > min_temperature
    next_solution = current_solution.dup
    i, j = rand(current_solution.size), rand(current_solution.size)
    next_solution[i], next_solution[j] = next_solution[j], next_solution[i]
    next_cost = yield(next_solution)

    if next_cost < current_cost or Math.exp((current_cost - next_cost) / temperature) > rand
      current_solution = next_solution
      current_cost = next_cost
    end

    temperature *= alpha
  end

  current_solution
end

# Example usage:
input_array = gets.chomp.split.map(&:to_i)
optimized_array = simulated_annealing(input_array) { |arr| arr.reduce(0) { |sum, num| sum + num ** 2 } }
puts "Optimized Array using Simulated Annealing: #{optimized_array}"

File: summary/elixir/22_annealing.exs
defmodule SimulatedAnnealing do
  def optimize(initial_solution, cost_function) do
    current_solution = initial_solution
    current_cost = cost_function.(current_solution)
    temperature = 100.0
    min_temperature = 0.01
    alpha = 0.9

    optimize_loop(current_solution, current_cost, temperature, min_temperature, alpha, cost_function)
  end

  defp optimize_loop(current_solution, current_cost, temperature, min_temperature, alpha, cost_function) do
    if temperature > min_temperature do
      next_solution = generate_neighbor(current_solution)
      next_cost = cost_function.(next_solution)

      {new_solution, new_cost} =
        if should_accept?(current_cost, next_cost, temperature) do
          {next_solution, next_cost}
        else
          {current_solution, current_cost}
        end

      new_temperature = temperature * alpha
      optimize_loop(new_solution, new_cost, new_temperature, min_temperature, alpha, cost_function)
    else
      current_solution
    end
  end

  defp generate_neighbor(solution) do
    i = :rand.uniform(length(solution)) - 1
    j = :rand.uniform(length(solution)) - 1
    List.replace_at(solution, i, Enum.at(solution, j))
    |> List.replace_at(j, Enum.at(solution, i))
  end

  defp should_accept?(current_cost, next_cost, temperature) do
    next_cost < current_cost or :math.exp((current_cost - next_cost) / temperature) > :rand.uniform()
  end
end

# Example usage:
input_array = IO.gets("Enter numbers separated by spaces: ")
              |> String.trim()
              |> String.split()
              |> Enum.map(&String.to_integer/1)

cost_function = fn arr -> Enum.reduce(arr, 0, fn num, sum -> sum + :math.pow(num, 2) end) end

optimized_array = SimulatedAnnealing.optimize(input_array, cost_function)
IO.puts("Optimized Array using Simulated Annealing: #{inspect(optimized_array)}")
